standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/PersistenceManagerProvider.java

```diff
@@ -302,23 +307,33 @@ private static PersistenceManagerFactory initPMF(Configuration conf, boolean for
       pmf = JDOHelper.getPersistenceManagerFactory(dsProp);
     } else {
       String sourceName = forCompactor ? "objectstore-compactor" : "objectstore";
+      DataSource ds = null, ds2 = null;
       try (DataSourceProvider.DataSourceNameConfigurator configurator =
                new DataSourceProvider.DataSourceNameConfigurator(conf, sourceName)) {
-        DataSource ds = (maxPoolSize > 0) ? dsp.create(conf, maxPoolSize) : dsp.create(conf);
+        ds = (maxPoolSize > 0) ? dsp.create(conf, maxPoolSize) : dsp.create(conf);
         // The secondary connection factory is used for schema generation, and for value generation operations.
         // We should use a different pool for the secondary connection factory to avoid resource starvation.
         // Since DataNucleus uses locks for schema generation and value generation, 2 connections should be sufficient.
         configurator.resetName("objectstore-secondary");
-        DataSource ds2 = forCompactor ? ds : dsp.create(conf, /* maxPoolSize */ 2);
+        ds2 = forCompactor ? ds : dsp.create(conf, /* maxPoolSize */ 2);
         dsProp.put(PropertyNames.PROPERTY_CONNECTION_FACTORY, ds);
         dsProp.put(PropertyNames.PROPERTY_CONNECTION_FACTORY2, ds2);
+        databaseProduct = DatabaseProduct.determineDatabaseProduct(ds, conf);
         dsProp.put(ConfVars.MANAGER_FACTORY_CLASS.getVarname(),
             "org.datanucleus.api.jdo.JDOPersistenceManagerFactory");
         pmf = JDOHelper.getPersistenceManagerFactory(dsProp);
       } catch (SQLException e) {
         LOG.warn("Could not create PersistenceManagerFactory using "
             + "connection pool properties, will fall back", e);
         pmf = JDOHelper.getPersistenceManagerFactory(prop);
+      } finally {
+        if (pmf == null && ds instanceof AutoCloseable) {
+          try (AutoCloseable close1 = (AutoCloseable) ds;
```

Comment:
why is this magic, can't we call close directly?

Reply:
The `DataSource` doesn't have the `close` directly to free the resource. For `HikariDataSource` and `PoolingDataSource`, they are implemented both the `AutoCloseable` and `DataSource`, so here need to cast the `ds` to the `AutoCloseable` explicitly.

Reply:
can't we refactor to use try-with-resources?
````
try (DataSource ds = create.., ds2 = create..) {
....
}
````

Reply:
can we add a `close` method to `DataSourceProvider`
````
  static void close(DataSource ds) {
    if (ds instanceof AutoCloseable) {
      try {
        ((AutoCloseable)ds).close();
      } catch (Exception e) {
        LOG.warn("Failed to close the DataSource", e);
      }
      }
    }
  }
````

Reply:
I would like to keep it, we don't introduce anything other than `close` in this method, and we have no chance to reuse it elsewhere.

Reply:
c3p0 provides the same method for handling pooled DS. why use try-with-resources if the block body is empty??

Reply:
The `close` method in `AutoCloseable` declared throwing the generic exception, on the caller, we should catch it anyhow. The issues mentioned in the link is because the `pmf` might be null, but this should never happen, as the exception should be thrown during create the `pmf`, the code cannot reach here in this case, but I also fixed it in the new commit.

Reply:
Nested blocks of code should not be left empty [java:S108](https://sonarcloud.io/organizations/apache/rules?open=java%3AS108&rule_key=java%3AS108)

Reply:
Added a debug log in this block